integration/kotlinx-coroutines-guava/src/ListenableFuture.kt

```diff
@@ -333,130 +329,161 @@ private class ListenableFutureCoroutine<T>(
+     * This arrangement means that [jobToCancel] _might not successfully cancel_, if the race resolves
+     * in a particular way. [jobToCancel] may also be in its "cancelling" state while this
      * ListenableFuture is complete and cancelled.
-     *
-     * [OuterFuture] collaborates with this class to present a more cohesive picture and ensure
-     * that certain combinations of cancelled/cancelling states can't be observed.
      */
     override fun cancel(mayInterruptIfRunning: Boolean): Boolean {
-        return if (super.cancel(mayInterruptIfRunning)) {
-            deferred.cancel()
+        // TODO: call jobToCancel.cancel() _before_ running the listeners.
+        //  `auxFuture.cancel()` will execute auxFuture's listeners. This delays cancellation of
+        //  `jobToCancel` until after auxFuture's listeners have already run.
+        //  Consider moving `jobToCancel.cancel()` into [AbstractFuture.afterDone] when the API is finalized.
+        return if (auxFuture.cancel(mayInterruptIfRunning)) {
+            jobToCancel.cancel()
             true
         } else {
             false
         }
     }
+
+    override fun toString(): String = buildString {
+        append(super.toString())
+        append("[status=")
+        if (isDone) {
+            try {
+                when (val result = Uninterruptibles.getUninterruptibly(auxFuture)) {
+                    is Result.Success -> append("SUCCESS, result=[${result.value}")
+                    is Result.Cancellation -> append("CANCELLED, cause=[${result.exception}]")
+                }
+            } catch (e: CancellationException) {
+                // `this` future was cancelled by `Future.cancel`. In this case there's no cause or message.
+                append("CANCELLED")
+            } catch (e: ExecutionException) {
+                append("FAILURE, cause=[${e.cause}]")
+            } catch (t: Throwable) {
+                // Violation of Future's contract, should never happen.
+                append("UNKNOWN, cause=[${t.javaClass} thrown from get()]")
+            }
+        } else {
+            append("PENDING, delegate=[$auxFuture]")
+        }
+    }
+}
+
+/**
+ * Represents the result of a `Coroutine` in the sense of [Future].
+ *
+ * - When coroutine [isCompleted][Job.isCompleted] successfully, its result is wrapped into [Success].
+ * - When coroutine [isCompleted][Job.isCompleted] exceptionally, meaning [isCancelled]
+ *   returns `true`. According to [Future]'s contract, the exception can be of two types:
+ *     - [CancellationException] if the coroutine was _cancelled normally_, meaning it threw [CancellationException].
+ *       In this case the exception is wrapped into [Result.Cancellation].
+ *     - [ExecutionException] with the original [Throwable] in its [cause][ExecutionException.cause] otherwise.
+ *       This case is handled separately. See [JobListenableFuture.completeExceptionally].
+ */
+private sealed class Result<out T> {
```

Comment:
Thanks for the explanation!
Would suggest getting rid of the `Result` clas tho: `Result.success` means a different thing in Kotlin (name clash with a popular stdlib class is not a great thing to have) and we also have to allocate on successful completion.

We can use raw value for the successful results and a `Cancelled(exception)` wrapper for cancellation exception. `isCancelled` then can be replaced with `val Any.is[Future]Cancelled() get() = this is Cancelled`